### PR TITLE
Browser module for `sort_by_line_nos()`  for an array of `ASTNode`s  + (Hidden Test Fix)

### PR DESCRIPTION
**Pull request by TilakMaddy**
This method helps in sorting the given array of nodes in the same order in which they appear in the source code.

This can help in checking for patterns like CEI, Re-entrancy, etc

------
Optional  read

`sort_by_` prefix was chosen because in Rust offers methods like `sort_by` , `sort_by_key`, etc so I thought this  would be the most intuitive to guess

### CODE REVIEW

File: aderyn_core/src/detect/experimental/parent_chain.rs [hunk 1]
```diff
@@ -79,7 +84,7 @@ mod parent_chain_demo_tests {
             "../tests/contract-playground/out/ParentChainContract.sol/ParentChainContract.json",
         );
 
-        let mut detector = ImmediateParentDemonstrator::default();
+        let mut detector = ParentChainDemonstrator::default();
```
Comment (TilakMaddy) [inline aderyn_core/src/detect/experimental/parent_chain.rs:87]: We mistakenly used some other detector demonstrator in this file. So I changed it now :)

File: aderyn_core/src/detect/experimental/parent_chain.rs [hunk 2]
```diff
@@ -26,20 +26,25 @@ In ParentChainContract.sol, there is only 1 assignment done. The goal is to capt
...
+                if let ASTNode::Block(block) = parent_chain[3] {
                     capture!(self, context, block);
                 }
             }
+
+            if let Some(mut parent_chain) = assignment.parent_chain(context) {
+                let sorted_chain = parent_chain.sort_by_line_nos(context).unwrap();
+                parent_chain.reverse();
+                assert_eq!(sorted_chain, parent_chain);
+            }
```
Comment (TilakMaddy) [inline aderyn_core/src/detect/experimental/parent_chain.rs:47]: Here you can see that the `parent_chain` when reversed becomes equal to the sorted version of itself. !!

File: aderyn_core/src/context/browser/sort_nodes.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
+use std::collections::BTreeSet;
+
+use crate::context::workspace_context::{ASTNode, WorkspaceContext};
+
+pub trait SortNodeReferencesToSequence<'a> {
+    fn sort_by_line_nos(self, context: &'a WorkspaceContext) -> Option<Vec<&'a ASTNode>>;
+}
```
Comment (TilakMaddy) [inline aderyn_core/src/context/browser/sort_nodes.rs:7]: We will use this more

Comment (TilakMaddy) [reply aderyn_core/src/context/browser/sort_nodes.rs:7]: (guessing)

File: aderyn_core/src/context/browser/sort_nodes.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+pub trait SortOwnedNodesToSequence<'a> {
+    fn sort_by_line_nos(self, context: &'a WorkspaceContext) -> Option<Vec<ASTNode>>;
+}
+
```
[the remark this answers is not in the document]
Comment (TilakMaddy) [reply aderyn_core/src/context/browser/sort_nodes.rs:12]: (guessing)

File: aderyn_core/src/context/browser/sort_nodes.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        return None;
+    }
+
+    // Now sort them
+    let mut nodes = nodes.to_vec();
+    nodes.sort_by(|a, b| {
+        context
+            .get_relative_location_of_nodes(a.id().unwrap(), b.id().unwrap())
+            .unwrap()
+    });
```
Comment (TilakMaddy) [inline aderyn_core/src/context/browser/sort_nodes.rs:58]: heart of the Logic

File: aderyn_core/src/detect/experimental/parent_chain.rs (hunk 2 above)
Comment (TilakMaddy) [reply aderyn_core/src/detect/experimental/parent_chain.rs:47]: (Example)

**Comment by TilakMaddy** [comment]
Follow up 🤞 (if this PR is merged, it will help me write another browser module that can help in detecting missing natspec, notice, dev, etc) 😉

**Comment by alexroan** [comment]
@alexroan Review this next

**Review by alexroan — APPROVED** [review]

**Pull request merged**